File: 201903/5.cpp
#include<bits/stdc++.h>

using namespace std;

#define MAXSIZE 10010
#define MAXNUM 0x7f7f7f7f
int graph[MAXSIZE][MAXSIZE];
bool types[MAXSIZE];
int targets[MAXSIZE];


void output(int n){
    for(int i=1;i<=n;i++){
        for(int j=1;j<=n;j++){
            printf("%d ",graph[i][j]);
        }
        printf("\n");
    }
    printf("\n");
}


int main(){
    
    
    int n,m,k;scanf("%d%d%d",&n,&m,&k);
    for(int i=1;i<=n;i++) scanf("%d",&types[i]);

    // for(int i=1;i<=n;i++)
    //     memset(graph[i],0x7f,sizeof(int)*(n+1));
    // for(int i=1;i<=n;i++) graph[i][i]=0;
    
    for(int i=1; i<=n; i++)
    {
        for(int j=1; j<=n; j++)
        {
            if(i==j)graph[i][j]=0;
            else graph[i][j]=MAXNUM;
        }
    }

    for(int i=0;i<m;i++){
        int u,v,w;scanf("%d%d%d",&u,&v,&w);
        if(u==v)continue;
        if(w<graph[u][v]){
            graph[u][v]=w;
            graph[v][u]=w;
        }
    }

    // output(n);

    for(int i=1;i<=n;i++)
        for(int j=1;j<=n;j++)
            for(int k=1;k<=n;k++)
                if(graph[j][i]<MAXNUM&&graph[i][k]<MAXNUM)
                    graph[j][k]=min(graph[j][k],graph[j][i]+graph[i][k]);

    // output(n);

    for(int i=1;i<=n;i++){
        memset(targets,0,(n+1)*sizeof(int));
        int top=0;
        for(int j=1;j<=n;j++){
            if(graph[i][j]<MAXNUM&&types[j]){
                targets[top++]=graph[i][j];
            }
        }
        sort(targets,targets+top);
        int count=0;
        for(int i=0;i<k&&i<top;i++){
            count+=targets[i];
        }
        printf("%d\n",count);
    }

}

File: 201903/5.py
n,m,kk=list(map(int,input().split()[:3]))
types = [0]+list(map(int,input().split()[:n]))

from math import inf

graph = [[inf for _ in range(n+1)] for _ in range(n+1)]

for i in range(1,n+1):graph[i][i]=0

for _ in range(m):
    u,v,w=list(map(int,input().split()))
    if u==v:continue
    graph[u][v]=graph[v][u]=min(w,graph[u][v])


for i in range(1,n+1):
    for j in range(1,n+1):
        for k in range(1,n+1):
            graph[j][k]=min(graph[j][k],graph[j][i]+graph[i][k])

##for i in range(1,n+1):
##    print(' '.join([str(j) for j in graph[i][1:]]))

for i in range(1,n+1):
    targets = [graph[i][j] for j in range(1,n+1) if types[j] and graph[i][j]<inf]
    targets.sort()
##    print(targets[:kk])
    if len(targets)<=kk:
        print(sum(targets))
    else:
        print(sum(targets[:kk]))
                             
    
    
